pkg/koordlet/runtimehooks/hooks/coresched/core_sched.go

```diff
@@ -0,0 +1,478 @@
+//  2. Add: Add a new cookie for a core sched group for a container, and add a new entry into the cache, ref count = 1.
+//  3. Assign: Assign a cookie of an existing core sched group for a container, and increase cookie's ref count in cache.
+//     The cookies of sibling PIDs (i.e. the PIDs of the same core sched group) will be fetched in the Assign. If all
+//     the cookies of sibling PIDs are default or invalid, the Assign will fall back to Add.
+//  4. Clear: Clear a cookie of an existing core sched group for a container (reset to default cookie 0), and the
+//     cookie's reference count in the cache is decreased. The cache entry will be removed when the ref count is no
+//     larger than zero.
+//
+// If multiple non-default cookies are assigned to existing containers of the same group, the firstly-created and
+// available cookie will be retained and the PIDs of others will be moved to the former.
+// NOTE: The agent itself should be set the default cookie. It can be excluded by setting QoS to SYSTEM.
+func (p *Plugin) SetContainerCookie(proto protocol.HooksProtocol) error {
+	containerCtx := proto.(*protocol.ContainerContext)
+	if containerCtx == nil {
+		return fmt.Errorf("container protocol is nil for plugin %s", name)
+	}
+	if !util.IsValidContainerCgroupDir(containerCtx.Request.CgroupParent) {
+		return fmt.Errorf("invalid container cgroup parent %s for plugin %s", containerCtx.Request.CgroupParent, name)
+	}
+
+	podUID := containerCtx.Request.PodMeta.UID
+	// only process sandbox container or container has valid ID
+	if len(podUID) <= 0 || len(containerCtx.Request.ContainerMeta.ID) <= 0 {
+		return fmt.Errorf("invalid container ID for plugin %s, pod UID %s, container ID %s",
+			name, podUID, containerCtx.Request.ContainerMeta.ID)
+	}
+
+	containerUID := p.getContainerUID(podUID, containerCtx.Request.ContainerMeta.ID)
+	podAnnotations := containerCtx.Request.PodAnnotations
+	podLabels := containerCtx.Request.PodLabels
+	podMetaName := containerCtx.Request.PodMeta.String()
+	containerName := containerCtx.Request.ContainerMeta.Name
+	podKubeQOS := util.GetKubeQoSByCgroupParent(containerCtx.Request.CgroupParent)
+
+	if !p.rule.IsInited() || !p.IsCacheInited() {
+		klog.V(5).Infof("plugin %s has not been inited, rule inited %v, cache inited %v, aborted to set cookie for container %s/%s",
+			name, p.rule.IsInited(), p.IsCacheInited(), podMetaName, containerName)
+		return nil
+	}
+
+	isEnabled, groupID := p.getPodEnabledAndGroup(podAnnotations, podLabels, podKubeQOS, podUID)
+	klog.V(6).Infof("manage cookie for container %s/%s, isEnabled %v, groupID %s",
+		podMetaName, containerName, isEnabled, groupID)
+
+	lastGroupID, lastCookieEntry, cookieEntry := p.getCookieCacheForContainer(groupID, containerUID)
+
+	// expect enabled
+	// 1. disabled -> enabled: Add or Assign.
+	// 2. keep enabled: Check the differences of cookie, group ID and the PIDs, and do Assign.
+	if isEnabled {
```

Comment:
this is a BIG if/else, could it be splited?

Reply:
done